### DISCUSSION TITLE
Bitmap drawing Elements missing

### DISCUSSION THREAD

**Discussion by adaumann** [AmigaPorts/ACE#238] [Ideas]
I would Like to port my C64 Game Orbix to the Amiga and add some more gameplay elements. This game is a pachinko/breakout style game like Peggle. The game is written in C and should not be to complicated, but I need some graphics library to draw Triangels/polygons (filled with patterns), filled circles (with patterns). All Elements should be clipped. This is missing, just line drawing I found. I‘m not very familar with Amiga Programming…

 Link to the C64 Version: https://fleischgemuese.itch.io/orbix

**Comment by tehKaiN** [comment]
Hi, thanks for your interest in ACE!

Indeed, support for such features is limited, because there was no prior need for those kinds of functions and ACE is developed in a Just In Time manner - new stuff that's developed for ACE-based games slowly trickles into the main library code after it's battle-tested and stable/universal enough.

Apart from line-drawing, you will need something for filling polygons, and Amiga blitter can accomodate that, but I'm not sure if patterned fill is something straightforward to do. I'll have to ask around.

On the other hand, if levels are predefined, perhaps it'd be plausible to keep level graphics pre-drawn (either in tiles or monolithic bitmap)? I see that you change patterns on objects as you hit them, but I also see that on C64 version it happens with limited granularity - perhaps you could store those parts in separate bitmaps and draw them on background as you hit them?

**Reply by adaumann** [reply]
I think I need the polygon drawing with patterns similar to the C64 version, in the C64 all levels are calculated during runtime and then drawn. Also the patterns might change, thats correct. Everything prerendering is too much effort and breaks the possibility to port the game. In a Amiga Version probably the polygons might move too or somehow animated.

In the C64 version I used the bitmap library from the OscarC compiler which is quite useful. 

Maybe someone else has an idea.

**Reply by tehKaiN** [reply]
Ok so I've gave it a spin and it's not that simple.

The blitter fill feature is done such that it can't be used to do the auto-pattern along the way, so what you'd have to do is to have bitmap A for mask, B for repeating pattern at least of size of final object, and a destination (D) bitmap, then:

- draw a shape using lines in bitmap A
- fill the shape with solid color in bitmap A
- use the bitmap A as mask and B as source and draw it using masked operation onto bitmap D

at this point, unless shapes are changing at runtime, it would be way way faster (line/fill operations are quite costly, thus rarely used) and easier to have those mask shapes precalculated and supplied as bitmaps. You could precalculate them in your game of course, but for quick and dirty prototyping you could supply them from file for now.

I'll add the solid fill function soon, as I have the code running but not really production-ready right now.

**Reply by adaumann** [reply]
Sounds very interesting,probably I have a chance to bring it to the Amiga. If you like I could test it,

**Reply by tehKaiN** [reply]
I've just pushed some changes to ACE's main branch:

- there's now blitLinePlane() function, which allows to draw lines on a single bitplane of the bitmap, which is considerably faster than blitLine(), and is basically what every line-based demo effect uses, as well as will be helpful with preparing masks
- I've added blitFillAligned() which fills the shape contained within specified rectangle with solid color on a given bitplane - complements blitLinePlane() nicely
- there's also a better example in ACE's showcase project on how to draw and fill shapes with blitter.

I still think that the better course of action would be using pre-supplied masks, but who am I to stop you. ;)

If you'd like a more direct ACE-related hotline/helpdesk/brainstorming, and you're into discord thingy, we have a dedicated channel on AmigaDev: https://discord.gg/Dfba7Yb

Good luck with your project and lemme know if you need anything else!

**Reply by adaumann** [reply]
Wow, that was quick. I‘m a bit busy but it is a good starting point to continue. Thank you
